pandas/_libs/algos_rank_helper.pxi.in

```diff
@@ -91,14 +93,24 @@ def rank_1d_{{dtype}}(object in_arr, ties_method='average', ascending=True,
     {{endif}}
 
     np.putmask(values, mask, nan_value)
+    {{else}}
+    mask = np.zeros(shape=len(values), dtype=bool)
     {{endif}}
 
     n = len(values)
     ranks = np.empty(n, dtype='f8')
 
     {{if dtype == 'object'}}
     try:
-        _as = values.argsort()
+        # lexsort on object array will raise TypeError for numpy version 
+        # earlier than 1.11.0
```

Comment:
why are you adding all of this?

Reply:
I want to keep nan values at the every end or at the every beginning. Otherwise sum_ranks and dups will be messed up when real +/- infs are intertwined with replaced infs (originally Nans )
```
            if tiebreak == TIEBREAK_AVERAGE:
                for j in range(i - dups + 1, i + 1):
                    ranks[argsorted[j]] = sum_ranks / dups
```
It is hard to let j skip the index corresponding to nan value, and to subtract extra counts from sum_ranks and dups at the same time. So I try to separate nans when I calculate _as. 
For float and int dtype, it is easy to achieve this by a double sort.
```
_as = np.lexsort(keys=(mask, values))
```
But numpy does not support lexsort on object array before version 1.11.0. 
https://github.com/numpy/numpy/pull/6312
Then I tried argsort with an order argument
```
_dt = [('values', 'object'), ('mask', 'bool')]
_values = np.asarray(list(zip(mask, values)), dtype=_dt)
_as = np.argsort(_values, kind='mergesort', order=('values', 'mask'))
```
Somehow, it is not working properly on Infinity, NegInfinity object. As a result, I sort it by mask first. The downside is I need to specify the condition on whether to reverse the sign of mask so as to control whether put Nans at the beginning or the end.

Reply:
ok can you make this a try/except/else to isolate the thing which is actually raising

Reply:
Infinity and NegInfinity might work if you add missing methods (might be a comparison operator eg dunder lt)

Reply:
Do you mean nesting the try/except clause?
```
    {{if dtype == 'object'}}
    try:
        # lexsort on object array will raise TypeError for numpy version 
        # earlier than 1.11.0
        _as = np.lexsort(keys=(mask, values))
    except TypeError:
        try:
            _dt = [('values', 'O'), ('mask', '?')]
            _values = np.asarray(list(zip(values, mask)), dtype=_dt)
            _as = np.argsort(_values, kind='mergesort', order=('values', 'mask'))
        except TypeError:
            if not retry:
                raise
            valid_locs = (~mask).nonzero()[0]
            ranks.put(valid_locs, rank_1d_object(values.take(valid_locs), 0,
                                                 ties_method=ties_method,
                                                 ascending=ascending))
            np.putmask(ranks, mask, np.nan)
            return ranks
    {{else}}
```

Reply:
no I mean try/except/else. you should put minimal amounts in the try portion; else is evaluation if no exception is raised. nested try/excepts are way harder to read.

Reply:
> Somehow, it is not working properly on Infinity, NegInfinity object. As a result, I sort it by mask first. The downside is I need to specify the condition on whether to reverse the sign of mask so as to control whether put Nans at the beginning or the end.

looks like you fixed Infinity, doesn't this original strategy work (and is across versions)?

Reply:
Sorry, I still don't know how to do it. (I get it , what try/except/else does)
When I am trying,
```
_as = np.lexsort(keys=(mask, values))
```
I will be expecting two kinds of errors when sorting an object array. One is due to lexsort is buggy in certain numpy versions, which can be fixed by argsort with order argument. Another is due to incomparable types in the array like comparing a str with an int. But they all raise same TypeError.  I don't know how to isolate them.
Moreover, if I use argsort with order argument to handle the version issue, it still might fail due to incomparable types within the array and raise another TypeError
What is the Pythonic way to fix it other they nested try/except ?